sam/contrib/postgres/tools.py

```diff
@@ -0,0 +1,36 @@
+import json
+import logging
+
+import psycopg
+
+from . import config
+
+logger = logging.getLogger(__name__)
+
+
+def fetch_all(query: str, _context=None) -> str:
+    """Fetch data from a PostgreSQL database using a SELECT query and return it as JSON.
+
+    Args:
+        query: SQL query to execute. It should be a SELECT statement.
+    """
+    with (
+        psycopg.connect(
+            dbname=config.DATABASE,
+            user=config.USERNAME,
+            password=config.PASSWORD,
+            host=config.HOSTNAME,
+            port=config.PORT,
+        ) as conn,
```

Comment:
What about we add `default_transaction_read_only=on` here?

https://www.postgresql.org/docs/current/runtime-config-client.html#GUC-DEFAULT-TRANSACTION-READ-ONLY


```suggestion
        psycopg.connect(
            dbname=config.DATABASE,
            user=config.USERNAME,
            password=config.PASSWORD,
            host=config.HOSTNAME,
            port=config.PORT,
            options=f"-c default_transaction_read_only=on",            
        ) as conn,
```